<a href="https://colab.research.google.com/github/maximepeg/BDF_notebooks/blob/main/BDF_12_Exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#00 - Configuration of Apache Spark on Collaboratory


###Installing Java, Spark, and Findspark


---


This code installs Apache Spark 3.0.1, Java 8, and [Findspark](https://github.com/minrk/findspark), a library that makes it easy for Python to find Spark.

In [ ]:
!apt-get update
!apt-get install openjdk-8-jdk-headless -qq > /dev/null
!wget  http://apache.osuosl.org/spark/spark-3.3.1/spark-3.3.1-bin-hadoop3.tgz  
!tar xf spark-3.3.1-bin-hadoop3.tgz  
!rm spark-3.3.1-bin-hadoop3.tgz    
!pip install -q findspark

Get:1 https://cloud.r-project.org/bin/linux/ubuntu focal-cran40/ InRelease [3,622 B]
Ign:2 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu2004/x86_64  InRelease
Hit:3 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64  InRelease
Hit:4 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu2004/x86_64  Release
Hit:5 http://archive.ubuntu.com/ubuntu focal InRelease
Get:6 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu focal InRelease [18.1 kB]
Get:7 http://security.ubuntu.com/ubuntu focal-security InRelease [114 kB]
Get:8 http://archive.ubuntu.com/ubuntu focal-updates InRelease [114 kB]
Hit:9 http://ppa.launchpad.net/cran/libgit2/ubuntu focal InRelease
Get:10 http://archive.ubuntu.com/ubuntu focal-backports InRelease [108 kB]
Hit:11 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu focal InRelease
Hit:12 http://ppa.launchpad.net/graphics-drivers/ppa/ubuntu focal InRelease
Get:14 http://ppa.launchpad.net/c2d4u.team/

### Set Environment Variables
Set the locations where Spark and Java are installed.

In [ ]:
import os
os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-8-openjdk-amd64"
os.environ["SPARK_HOME"] = "/content/spark/"
os.environ["DRIVE_DATA"] = "/content/gdrive/My Drive/Enseignement/2022-2023/ING3/HPDA/BigDataFrameworks/data/"

!rm /content/spark
!ln -s /content/spark-3.3.1-bin-hadoop3 /content/spark
!export SPARK_HOME=/content/spark
!export PATH=$PATH:$SPARK_HOME/bin:$SPARK_HOME/sbin
!echo $SPARK_HOME
!env |grep  "DRIVE_DATA"

rm: cannot remove '/content/spark': No such file or directory
/content/spark/
DRIVE_DATA=/content/gdrive/My Drive/Enseignement/2022-2023/ING3/HPDA/BigDataFrameworks/data/


### Start a SparkSession
This will start a local Spark session.

In [ ]:
!python -V

import findspark
findspark.init()

from pyspark import SparkContext
sc = SparkContext.getOrCreate()

# Example: shows the PySpark version
print("PySpark version {0}".format(sc.version))

# Example: parallelise an array and show the 2 first elements
sc.parallelize([2, 3, 4, 5, 6]).cache().take(2)

Python 3.8.10
PySpark version 3.3.1


[2, 3]

In [ ]:
from pyspark.sql import SparkSession
# We create a SparkSession object (or we retrieve it if it is already created)
spark = SparkSession \
.builder \
.appName("My application") \
.config("spark.some.config.option", "some-value") \
.master("local[4]") \
.getOrCreate()
# We get the SparkContext
sc = spark.sparkContext

In [ ]:
# Mount Google Drive
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive



---


# 12 - Exercises. Final assignment

## Exercise 12.1

Let us extract information from the cite75_99.txt and apat63_99.txt files. Write a script that performs the following operations:

1. From the cite75_99.txt file, obtain the number of citations received by each patent. You must produce a DataFrame with the following format:

| PatentNum | ncitations |
|-----------|------------|
| 3060453   |    3       |
| 3390168   |    6       |
| 3626542   |   18       | 
| 3611507   |    5       |
| 3000113   |    4       |


2. From the apat63_99.txt file, create a DataFrame to show the patent number, its country and the patent year, discarding the rest of fields in the file. The DataFrame produced must have the following format:

|PatentNum |country|Year |
|----------|-------|-----|
| 3070801  | BE    | 1963| 
| 3070802  | US    | 1963| 
| 3070803  | US    | 1963| 
| 3070804  | US    | 1963| 
| 3070805  | US    | 1963|

 
**Requirements**

 - Both DataFrames must be stored in Parquet format with gzip compression. Check the number of partitions of each DataFrame and the number of files gererated.

 - It is **strongly advised** to copy the files from your Drive to a temporal directory in the notebook virtual machine and unzip them there. This will reduce the execution times. See the cell below:
        

In [ ]:
!mkdir /tmp/data/
!cp "$DRIVE_DATA"apat63_99.txt.tar.bz2 "$DRIVE_DATA"cite75_99.txt.tar.bz2 /tmp/data
%cd /tmp/data
!ls
!tar -jxf apat63_99.txt.tar.bz2
!tar -jxf cite75_99.txt.tar.bz2
#!rm /tmp/data/*.tar.bz2
!ls
!pwd

mkdir: cannot create directory ‘/tmp/data/’: File exists
/tmp/data
apat63_99.txt  apat63_99.txt.tar.bz2  cite75_99.txt  cite75_99.txt.tar.bz2
apat63_99.txt  apat63_99.txt.tar.bz2  cite75_99.txt  cite75_99.txt.tar.bz2
/tmp/data


In [ ]:
df2 = spark.read.option("header",True).csv("/tmp/data/" + "apat63_99.txt")
df = spark.read.option("header",True).csv(os.environ["DRIVE_DATA"] + "cite75_99.txt")


In [ ]:
print(df2.show(10))
print(df.show(10))

+-------+-----+-----+-------+-------+-------+--------+-------+------+------+---+------+-----+--------+--------+-------+--------+--------+--------+--------+--------+--------+--------+
| PATENT|GYEAR|GDATE|APPYEAR|COUNTRY|POSTATE|ASSIGNEE|ASSCODE|CLAIMS|NCLASS|CAT|SUBCAT|CMADE|CRECEIVE|RATIOCIT|GENERAL|ORIGINAL|FWDAPLAG|BCKGTLAG|SELFCTUB|SELFCTLB|SECDUPBD|SECDLWBD|
+-------+-----+-----+-------+-------+-------+--------+-------+------+------+---+------+-----+--------+--------+-------+--------+--------+--------+--------+--------+--------+--------+
|3070801| 1963| 1096|   null|     BE|   null|    null|      1|  null|   269|  6|    69| null|       1|    null|      0|    null|    null|    null|    null|    null|    null|    null|
|3070802| 1963| 1096|   null|     US|     TX|    null|      1|  null|     2|  6|    63| null|       0|    null|   null|    null|    null|    null|    null|    null|    null|    null|
|3070803| 1963| 1096|   null|     US|     IL|    null|      1|  null|     2|  6|    6

In [ ]:
citation = df.groupBy("CITING").count().withColumnRenamed("count", "ncitations")
citation.show(10)

+-------+----------+
| CITING|ncitations|
+-------+----------+
|3858333|         3|
|3858342|         5|
|3858485|         4|
|3858770|         4|
|3858812|         4|
|3859258|         3|
|3859358|         2|
|3859718|         3|
|3859739|         5|
|3859806|         5|
+-------+----------+
only showing top 10 rows



In [ ]:
citation.write.mode('overwrite').parquet("/tmp/data/citations.parquet")
citation.rdd.getNumPartitions()

2

In [ ]:
patent_info = df2.select('PATENT', 'COUNTRY','GYEAR')
patent_info.show(10)

+-------+-------+-----+
| PATENT|COUNTRY|GYEAR|
+-------+-------+-----+
|3070801|     BE| 1963|
|3070802|     US| 1963|
|3070803|     US| 1963|
|3070804|     US| 1963|
|3070805|     US| 1963|
|3070806|     US| 1963|
|3070807|     US| 1963|
|3070808|     US| 1963|
|3070809|     US| 1963|
|3070810|     US| 1963|
+-------+-------+-----+
only showing top 10 rows



In [ ]:
patent_info.write.mode('overwrite').parquet("/tmp/data/patent.parquet")
patent_info.rdd.getNumPartitions()

2

In [ ]:
!ls /tmp/data/*.parquet

/tmp/data/citations.parquet:
part-00000-345a153d-4518-40eb-98d1-991b52c27260-c000.snappy.parquet  _SUCCESS
part-00001-345a153d-4518-40eb-98d1-991b52c27260-c000.snappy.parquet

/tmp/data/patent.parquet:
part-00000-3b0bc1a3-63a4-47dc-9fa7-d4e0c577d2ca-c000.snappy.parquet  _SUCCESS
part-00001-3b0bc1a3-63a4-47dc-9fa7-d4e0c577d2ca-c000.snappy.parquet


## Exercise 12.2

Write a code that, from the Parquet files created in the previous exercise, obtains for each country and for each year: the total number of patents, the total number of citations from those patents, the average number of citations and the maximum number of citations. Compute only those values in which there are any values in both files (*inner join*). In addition, each country must show its whole name, obtained from the *country_codes.txt* file. The final DataFrame must look like this one:


|Country            |Year|PatentsNum |TotalCitations|AvgCitations      |MaxCitations|
|-------------------|----|-----------|--------------|------------------|------------|
|Algeria            |1963|2          |7             |3.5               |4           |
|Algeria            |1968|1          |2             |2.0               |2           |
|Algeria            |1970|1          |2             |2.0               |2           |
|Algeria            |1972|1          |1             |1.0               |1           |
|Algeria            |1977|1          |2             |2.0               |2           |
|Andorra            |1987|1          |3             |3.0               |3           |
|Andorra            |1993|1          |1             |1.0               |1           |
|Andorra            |1998|1          |1             |1.0               |1           |
|Antigua and Barbuda|1978|1          |6             |6.0               |6           |
|Antigua and Barbuda|1979|1          |14            |14.0              |14          |
|Antigua and Barbuda|1991|1          |8             |8.0               |8           |
|Antigua and Barbuda|1994|1          |19            |19.0              |19          |
|Antigua and Barbuda|1995|2          |12            |6.0               |11          |
|Antigua and Barbuda|1996|2          |3             |1.5               |2           |
|Argentina          |1963|14         |35            |2.5               |7           |
|Argentina          |1964|20         |60            |3.0               |8           |
|Argentina          |1965|10         |35            |3.5               |10          |
|Argentina          |1966|16         |44            |2.75              |9           |
|Argentina          |1967|13         |60            |4.615384615384615 |14          |

**Requirements**

- The output DataFrame must be saved in a single CSV file, with a header and without any compression.


In [ ]:
!cp "$DRIVE_DATA"country_codes.txt /tmp/data/country_codes.txt
!ls /tmp/data


apat63_99.txt	       citations.parquet  cite75_99.txt.tar.bz2  patent.parquet
apat63_99.txt.tar.bz2  cite75_99.txt	  country_codes.txt


In [ ]:
countries = spark.read.options(header=False, delimiter="\t").csv("/tmp/data/country_codes.txt").withColumnRenamed("_c0", "COUNTRY_CODE").withColumnRenamed("_c1", "COUNTRY")
cit = spark.read.parquet("/tmp/data/citations.parquet").withColumnRenamed("CITING", "PATENT")
pat = spark.read.parquet("/tmp/data/patent.parquet").withColumnRenamed("COUNTRY", "COUNTRY_CODE")
print(pat.show(5))
print(cit.show(5))
print(countries.show(5))

+-------+------------+-----+
| PATENT|COUNTRY_CODE|GYEAR|
+-------+------------+-----+
|3070801|          BE| 1963|
|3070802|          US| 1963|
|3070803|          US| 1963|
|3070804|          US| 1963|
|3070805|          US| 1963|
+-------+------------+-----+
only showing top 5 rows

None
+-------+----------+
| PATENT|ncitations|
+-------+----------+
|3858333|         3|
|3858342|         5|
|3858485|         4|
|3858770|         4|
|3858812|         4|
+-------+----------+
only showing top 5 rows

None
+------------+--------------+
|COUNTRY_CODE|       COUNTRY|
+------------+--------------+
|          AF|   Afghanistan|
|          AX| Aland Islands|
|          AL|       Albania|
|          DZ|       Algeria|
|          AS|American Samoa|
+------------+--------------+
only showing top 5 rows

None


In [ ]:
from pyspark.sql.functions import sum,avg,max, count, countDistinct

In [ ]:
data_df = cit.join(pat, on="PATENT", how="inner").join(countries, on="COUNTRY_CODE", how="left").withColumnRenamed("GYEAR","YEAR")

full_patent_stats =data_df.groupBy("COUNTRY", "YEAR").agg(countDistinct("PATENT").alias("PatentNums"), 
                                                          sum("ncitations").alias("TotalCitations"), 
                                                          avg("ncitations").alias("AvgCitations"),
                                                          max("ncitations").alias("MaxCitations"))
full_patent_stats.show(5)

+--------------------+----+----------+--------------+------------------+------------+
|             COUNTRY|YEAR|PatentNums|TotalCitations|      AvgCitations|MaxCitations|
+--------------------+----+----------+--------------+------------------+------------+
|            Colombia|1978|         7|            24|3.4285714285714284|           7|
|           Australia|1979|       207|          1075| 5.193236714975845|          15|
|Serbia and Monten...|1980|        13|            73| 5.615384615384615|          13|
|              Canada|1980|      1055|          6020| 5.706161137440758|          32|
|Czechoslovakia (f...|1980|        53|           253| 4.773584905660377|          13|
+--------------------+----+----------+--------------+------------------+------------+
only showing top 5 rows



## Exercise 12.3

From the apat63_99.txt file, obtain the number of patents per country and year **using RDDs** (do not use DataFrames). The resulting RDD must be a key/value RDD in which the key is a country and the value a list of tuples. Each tuple will be composed of a year and the number of patents of the country during that year. In addition, the resulting RDD must be sorted by  the country code and, for each country, values must be sorted by year.

Example of output key/value entry:

    (u'PA', [(u'1963', 2), (u'1964', 2), (u'1965', 1), (u'1966', 1), (u'1970', 1), (u'1971', 1), (u'1972', 6), (u'1974', 3), (u'1975', 5), (u'1976', 3), (u'1977', 2), (u'1978', 2), (u'1980', 2), (u'1982', 1), (u'1983', 1), (u'1985', 2), (u'1986', 1), (u'1987', 2), (u'1988', 1), (u'1990', 1), (u'1991', 2), (u'1993', 1), (u'1995', 1), (u'1996', 1), (u'1999', 1)])

**Requirements:**

- You must remove the double quotation marks from the country code.
- Use 8 partitions to read the apat63_99.txt.bz2 file.


In [ ]:
tupled_rdd = spark.sparkContext.textFile("/tmp/data/apat63_99.txt", minPartitions=8).map(lambda x: x.split(",")).map(lambda x:((x[4].replace('"',''), x[1]), x[0] )).groupByKey().mapValues(len).map(lambda x: (x[0][0], (x[0][1],x[1]))).groupByKey().mapValues(list).sortBy(lambda x: (x[0], x[1][0]))

tupled_rdd.take(5)

[('AD', [('1999', 1), ('1995', 1), ('1998', 2), ('1993', 1), ('1987', 1)]),
 ('AE',
  [('1984', 2),
   ('1996', 1),
   ('1994', 1),
   ('1987', 1),
   ('1989', 2),
   ('1992', 1),
   ('1991', 2),
   ('1985', 2),
   ('1999', 3),
   ('1990', 1),
   ('1998', 1),
   ('1993', 1)]),
 ('AG',
  [('1995', 2),
   ('1999', 1),
   ('1991', 1),
   ('1979', 1),
   ('1994', 1),
   ('1996', 2),
   ('1978', 1)]),
 ('AI', [('1998', 1)]),
 ('AL', [('1999', 1)])]

## Exercise 12.4

From the Parquet files created in Exercise 12.1, create a DataFrame that gives the patent or patents with the higher number of citations per country and year, as well as the average of the number of citations per country and year, and the difference between the maximum and the average values. The resulting DataFrame should look like this:


|Country|Year|PatentNum|max  |average       |diff              |
|-------|----|---------|-----|--------------|------------------|
|AD     |1987|4688621  |3    |3.0           |0.0               |
|AD     |1993|5193231  |1    |1.0           |0.0               |
|AD     |1998|5765303  |1    |1.0           |0.0               |
|AE     |1984|4482959  |5    |5.0           |0.0               |
|AE     |1985|4554981  |14   |14.0          |0.0               |
|AE     |1987|4663181  |3    |3.0           |0.0               |
|AE     |1989|4805221  |7    |5.0           |2.0               |
|AE     |1990|4909321  |2    |2.0           |0.0               |
|AE     |1991|5004552  |3    |2.0           |1.0               |
|AE     |1992|5104556  |4    |4.0           |0.0               |
|AE     |1993|5181569  |8    |8.0           |0.0               |
|AE     |1996|5580125  |1    |1.0           |0.0               |
|AG     |1978|4126850  |6    |6.0           |0.0               |
|AG     |1979|4172981  |14   |14.0          |0.0               |
|AG     |1991|5013035  |8    |8.0           |0.0               |
|AG     |1994|5345071  |19   |19.0          |0.0               |
|AG     |1995|5457307  |11   |6.0           |5.0               |
|AG     |1996|5525786  |2    |1.5           |0.5               |

**Requirements:**

- The DataFrame must be sorted by country code and year.
- Do **NOT** replace the country code by its whole name.
- The output must be saved as a single CSV file, with a header and without any compression.


In [ ]:
cit = spark.read.parquet("/tmp/data/citations.parquet").withColumnRenamed("CITING", "PATENT")
pat = spark.read.parquet("/tmp/data/patent.parquet").withColumnRenamed("COUNTRY", "COUNTRY_CODE")
print(cit.show(5))
print(pat.show(5))

+-------+----------+
| PATENT|ncitations|
+-------+----------+
|3858333|         3|
|3858342|         5|
|3858485|         4|
|3858770|         4|
|3858812|         4|
+-------+----------+
only showing top 5 rows

None
+-------+------------+-----+
| PATENT|COUNTRY_CODE|GYEAR|
+-------+------------+-----+
|3070801|          BE| 1963|
|3070802|          US| 1963|
|3070803|          US| 1963|
|3070804|          US| 1963|
|3070805|          US| 1963|
+-------+------------+-----+
only showing top 5 rows

None


In [ ]:
from pyspark.sql.functions import max_by, abs
df_diff= cit.join(pat, on="PATENT", how="inner").groupBy("COUNTRY_CODE", "GYEAR").agg(max_by("PATENT", "ncitations").alias("patent"), 
                                                                                      max("ncitations").alias("max"),
                                                                                      avg("ncitations").alias("avg"),
                                                                                      abs(max("ncitations") - avg("ncitations")).alias("diff"))
df_diff.show(10)
df_diff.coalesce(1).write.csv("/tmp/data/patent_data_with_diff.csv")


+------------+-----+-------+---+----+----+
|COUNTRY_CODE|GYEAR| patent|max| avg|diff|
+------------+-----+-------+---+----+----+
|          AD| 1987|4688621|  9| 9.0| 0.0|
|          AD| 1993|5193231|  2| 2.0| 0.0|
|          AD| 1995|5478082|  8| 8.0| 0.0|
|          AD| 1998|5765303| 11| 7.0| 4.0|
|          AD| 1999|5894770|  4| 4.0| 0.0|
|          AE| 1984|4482959|  6| 5.5| 0.5|
|          AE| 1985|4532995| 11|10.5| 0.5|
|          AE| 1987|4663181|  4| 4.0| 0.0|
|          AE| 1989|4805221| 13|12.0| 1.0|
|          AE| 1990|4909321| 17|17.0| 0.0|
+------------+-----+-------+---+----+----+
only showing top 10 rows



!## Exercise 12.5

From the Parquet file with the (PatentNum,Country,Year) information from Exercise 12.1, create a DataFrame that shows the number of patents associated to each country per decade (understanding as a *decade* the years from 0 to 9; e.g. from 1970 to 1979). In addition, the DataFrame must show the increase or decrease of the number of patents per country and decade with respect to the previous decade. The resulting DataFrame must look like this:

|Country|Decade|PatentsNum|Diff|
|-------|------|----------|----|
|AD     |1980  |1         |0   |
|AD     |1990  |5         |4   |
|AE     |1980  |7         |0   |
|AE     |1990  |11        |4   |
|AG     |1970  |2         |0   |
|AG     |1990  |7         |5   |
|AI     |1990  |1         |0   |
|AL     |1990  |1         |0   |
|AM     |1990  |2         |0   |
|AN     |1970  |1         |0   |
|AN     |1980  |2         |1   |
|AN     |1990  |5         |3   |
|AR     |1960  |135       |0   |
|AR     |1970  |239       |104 |
|AR     |1980  |184       |-55 |
|AR     |1990  |292       |108 |

**Requirements**

- The DataFrame must be sorted by country code and year.
- Do **NOT** replace the country code by its whole name.
- The output must be saved as a single CSV file, with a header and without any compression.

In [ ]:
cit = spark.read.parquet("/tmp/data/citations.parquet").withColumnRenamed("CITING", "PATENT")
pat = spark.read.parquet("/tmp/data/patent.parquet").withColumnRenamed("COUNTRY", "COUNTRY_CODE")
print(cit.show(5))
print(pat.show(5))

+-------+----------+
| PATENT|ncitations|
+-------+----------+
|3858333|         3|
|3858342|         5|
|3858485|         4|
|3858770|         4|
|3858812|         4|
+-------+----------+
only showing top 5 rows

None
+-------+------------+-----+
| PATENT|COUNTRY_CODE|GYEAR|
+-------+------------+-----+
|3070801|          BE| 1963|
|3070802|          US| 1963|
|3070803|          US| 1963|
|3070804|          US| 1963|
|3070805|          US| 1963|
+-------+------------+-----+
only showing top 5 rows

None


In [ ]:
from pyspark.sql.functions import round, lag
from pyspark.sql.window import Window

w = Window.partitionBy("COUNTRY_CODE").orderBy("DECADE")
df_lag = cit.join(pat, on="PATENT", how="inner").withColumn("DECADE", round("GYEAR", -1)-10).groupBy("COUNTRY_CODE", "DECADE").agg(sum("ncitations").alias("patentNums"))
df_lag = df_lag.withColumn("lag", lag("patentNums", 0).over(w) - lag("patentNums", 1).over(w)).fillna(0)
df_lag.show(10)

+------------+------+----------+---+
|COUNTRY_CODE|DECADE|patentNums|lag|
+------------+------+----------+---+
|          AD|1980.0|        11|  0|
|          AD|1990.0|        26| 15|
|          AE|1970.0|        11|  0|
|          AE|1980.0|       115|104|
|          AE|1990.0|        52|-63|
|          AG|1970.0|        17|  0|
|          AG|1980.0|        13| -4|
|          AG|1990.0|        32| 19|
|          AI|1990.0|        11|  0|
|          AM|1990.0|        25|  0|
+------------+------+----------+---+
only showing top 10 rows



In [ ]:
df_lag.coalesce(1).write.csv("/tmp/data/patent_data_with_previous_year_comparison.csv")
